pkg/client/request.go

```diff
@@ -443,6 +449,252 @@ func (r *Request) Stream() (io.ReadCloser, error) {
+		go func() {
+			io.Copy(stdinW, stdin)
+			stdinR.Close()
+		}()
+		go func() {
+			cp("input", remoteStdin, stdinR)
+			remoteStdin.Close()
+			wg.Done()
+		}()
+	}
+
+	if doStdout {
+		wg.Add(1)
+		go cp("output", stdout, remoteStdout)
+	}
+
+	if doStderr {
+		wg.Add(1)
+		go cp("error", stderr, remoteStderr)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+// ForwardPorts formats and executes a port forwarding request. The connection will remain
+// open until the user interrupts the process.
+func (r *Request) ForwardPorts(ports []uint16) error {
+	if r.err != nil {
+		return r.err
+	}
+	req, err := http.NewRequest(r.verb, r.finalURL(), nil)
+	if err != nil {
+		return err
+	}
+
+	portForwarder := portForwarder{upgrader: spdy.NewConnectionUpgrader()}
+	return portForwarder.forward(req, ports)
+}
+
+// portForwarder knows how to listen for local connections and forward them to
+// a remote container via an upgraded HTTP request.
+type portForwarder struct {
+	upgrader   httpstream.ConnectionUpgrader
+	streamConn httpstream.Connection
+	listeners  []io.Closer
+}
+
+// forward dials the remote host specific in req, upgrades the request, starts
+// listeners for each port specified in ports, and forwards local connections
+// to the remote host via streams.
+func (pf *portForwarder) forward(req *http.Request, ports []uint16) error {
+	var err error
+	pf.streamConn, err = pf.upgrader.UpgradeRequest(req, httpstream.NoOpNewStreamHandler)
+	if err != nil {
+		return err
+	}
+	defer pf.streamConn.Close(1 * time.Second)
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
+
+	listenSuccess := false
+	for _, port := range ports {
+		err = pf.listenOnPort(port)
+		if err != nil {
+			glog.Warningf("Unable to listen on port %d: %v", port, err)
+		}
+		listenSuccess = true
+	}
+
+	if !listenSuccess {
+		return fmt.Errorf("Unable to listen on any of the requested ports: %v", ports)
+	}
+
+	// wait for interrupt
+	<-signals
+
+	// stop all listeners
+	for _, l := range pf.listeners {
+		l.Close()
+	}
+
+	return nil
+}
+
+// listenOnPort creates a new listener on port and waits for new connections
+// in the background.
+func (pf *portForwarder) listenOnPort(port uint16) error {
+	listener, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
+	if err != nil {
+		return err
+	}
+
+	pf.listeners = append(pf.listeners, listener)
+
+	go pf.waitForConnection(listener, port)
+
+	return nil
+}
+
+// waitForConnection waits for new connections to listener and handles them in
+// the background.
+func (pf *portForwarder) waitForConnection(listener net.Listener, port uint16) {
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			// TODO consider using something like https://github.com/hydrogen18/stoppableListener?
+			if !strings.Contains(strings.ToLower(err.Error()), "use of closed network connection") {
+				glog.Errorf("Error accepting connection on port %d: %v", port, err)
+			}
+			return
+		}
```

Comment:
I'm not sure what you mean - I'm logging everything except the error you get when another goroutine calls `listener.Close()`